### PR TITLE
Update compiler to consume local imports

### PR DESCRIPTION
**Pull request by tstirrat15**
## Description
Part of getting composable schemas implemented. This actually implements the traversal.

## Notes
This is intentionally incremental, and is not intended to be the release candidate. Future work includes:

* Import cycle tracking
* Only including requested imports (rather than all imports)

## Changes
* Plumb an `existingNames` value through the compiler so that it can track the definitions it's already seen
* Add file importer logic
* Implement import logic the translator
## Testing
Review. See that tests go green.

### CODE REVIEW

File: pkg/composableschemadsl/compiler/compiler.go [hunk 1]
```diff
@@ -52,6 +52,10 @@ func (cs CompiledSchema) SourcePositionToRunePosition(source input.Source, posit
 type config struct {
 	skipValidation   bool
 	objectTypePrefix *string
+	existingNames    []string
+	// In an import context, this is the folder containing
+	// the importing schema (as opposed to imported schemas)
+	sourceFolder string
```
Comment (tstirrat15) [inline pkg/composableschemadsl/compiler/compiler.go:58]: I don't really like how I need to plumb this through three levels in order to get it to the next compilation context. I'd welcome recommendations on a different way to structure this.

File: pkg/composableschemadsl/compiler/translator.go [hunk 2]
```diff
@@ -22,6 +23,8 @@ type translationContext struct {
 	mapper           input.PositionMapper
 	schemaString     string
 	skipValidate     bool
+	existingNames    []string
```
Comment (tstirrat15) [inline pkg/composableschemadsl/compiler/translator.go:26]: Should this be a `*mapz.Set` up and down the stack?

**Review by tstirrat15 — COMMENTED** [review]
A couple of thoughts

File: pkg/composableschemadsl/compiler/translator.go (hunk 2 above)
Comment (josephschorr) [reply pkg/composableschemadsl/compiler/translator.go:26]: Likely

File: pkg/composableschemadsl/compiler/compiler.go (hunk 1 above)
Comment (josephschorr) [reply pkg/composableschemadsl/compiler/compiler.go:58]: Why can't you put it on the translationContext?

File: pkg/composableschemadsl/compiler/importer.go [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+	var schemaBytes []byte
+	schemaBytes, err := os.ReadFile(filepath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read schema file: %w", err)
+	}
+	// TODO: do we want this sort of logging here?
+	log.Trace().Str("schema", string(schemaBytes)).Str("file", filepath).Msg("read schema from file")
+
+	compiled, err := Compile(InputSchema{
```
Comment (josephschorr) [inline pkg/composableschemadsl/compiler/importer.go:34]: If you call an internal compile, then this could be passed the translationContext directly and then you wouldn't need a different context here or the need to pass it into Compile

File: pkg/composableschemadsl/compiler/importer.go [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+	pathSegments []string
+	sourceFolder string
+	names        *mapz.Set[string]
+}
+
+const SchemaFileSuffix = ".zed"
+
+func ImportFile(importContext *ImportContext) (*CompiledSchema, error) {
+	relativeFilepath := constructFilePath(importContext.pathSegments)
+	filePath := path.Join(importContext.sourceFolder, relativeFilepath)
```
Comment (vroldanbet) [inline pkg/composableschemadsl/compiler/importer.go:25]: what happens if the imports are outside of the root context folder? is that supported, and if not, what kind of opinions/guidelines does it impose on the structure of a composable schema?

Comment (josephschorr) [reply pkg/composableschemadsl/compiler/importer.go:25]: No; the eventual goal will be to support this using git-based imports

File: .github/workflows/lint.yaml [hunk 5]
```diff
@@ -1,6 +1,6 @@
 ---
 name: "Lint"
-on: # yamllint disable-line rule:truthy
+on:  # yamllint disable-line rule:truthy
```
Comment (josephschorr) [inline .github/workflows/lint.yaml:3]: Why this change and below?

File: pkg/composableschemadsl/compiler/importer.go [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+
+type ImportContext struct {
+	pathSegments []string
+	sourceFolder string
+	names        *mapz.Set[string]
+}
+
+const SchemaFileSuffix = ".zed"
+
+func ImportFile(importContext *ImportContext) (*CompiledSchema, error) {
```
Comment (josephschorr) [inline pkg/composableschemadsl/compiler/importer.go:23]: Unless you see a need to export this, should we just keep this internal (i.e. lowercase)?

Comment (tstirrat15) [reply pkg/composableschemadsl/compiler/importer.go:23]: That works for me

File: .github/workflows/lint.yaml (hunk 5 above)
Comment (tstirrat15) [reply .github/workflows/lint.yaml:3]: `yamllint` wants two spaces before an inline comment and appeared to be failing the lint check as a result.

File: pkg/composableschemadsl/compiler/importer.go (hunk 4 above)
Comment (tstirrat15) [reply pkg/composableschemadsl/compiler/importer.go:25]: The biggest effect that I see is that you won't be able to do a relative import *within* the package that walks up a directory. this means that you couldn't have:

```
app1/app1.zed
app2/app2.zed
util.zed
```

and have both app files reference the util file. I have a feeling this will be a tripping point; it'll either have to be well-documented or something we change.

File: .github/workflows/lint.yaml (hunk 5 above)
Comment (josephschorr) [reply .github/workflows/lint.yaml:3]: Blech

File: pkg/composableschemadsl/compiler/compiler.go [hunk 7]
```diff
@@ -68,12 +72,21 @@ func AllowUnprefixedObjectType() ObjectPrefixOption {
 	return func(cfg *config) { cfg.objectTypePrefix = new(string) }
 }
 
+func SourceFolder(sourceFolder string) Option {
```
Comment (josephschorr) [inline pkg/composableschemadsl/compiler/compiler.go:75]: Add a doc comment explaining what this does. Probably good to add doc comments to the other exported functions too

File: pkg/composableschemadsl/compiler/importer.go [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+	"path"
+	"path/filepath"
+
+	"github.com/rs/zerolog/log"
+
+	"github.com/authzed/spicedb/pkg/composableschemadsl/input"
+	"github.com/authzed/spicedb/pkg/genutil/mapz"
+)
+
+type ImportContext struct {
```
Comment (josephschorr) [inline pkg/composableschemadsl/compiler/importer.go:15]: make this context unexported as well

File: pkg/composableschemadsl/compiler/importer.go [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+	pathSegments []string
+	sourceFolder string
+	names        *mapz.Set[string]
+}
+
+const SchemaFileSuffix = ".zed"
+
+func importFile(importContext *ImportContext) (*CompiledSchema, error) {
```
Comment (josephschorr) [inline pkg/composableschemadsl/compiler/importer.go:23]: why does this take in a `*` vs just the importContext struct itself?

File: pkg/composableschemadsl/compiler/importer.go [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+
+	newSourceFolder := filepath.Dir(filePath)
+
+	var schemaBytes []byte
+	schemaBytes, err := os.ReadFile(filePath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read schema file: %w", err)
+	}
+	// TODO: do we want this sort of logging here?
+	log.Trace().Str("schema", string(schemaBytes)).Str("file", filePath).Msg("read schema from file")
```
Comment (josephschorr) [inline pkg/composableschemadsl/compiler/importer.go:35]: yes; trace is good

File: pkg/composableschemadsl/compiler/importer.go [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+
+	compiled, err := compileImpl(InputSchema{
+		// TODO: should this point to the schema file? What is this for?
```
Comment (josephschorr) [inline pkg/composableschemadsl/compiler/importer.go:38]: yes; it gives the file path for error messaging

File: pkg/composableschemadsl/compiler/importer.go (hunk 9 above)
Comment (tstirrat15) [reply pkg/composableschemadsl/compiler/importer.go:23]: Mostly me not having built an intuition yet for when you would use one vs the other.

**Review by josephschorr — APPROVED** [review]
LGTM
